R/verb-fill.R

```diff
@@ -0,0 +1,169 @@
+# * hive: https://cwiki.apache.org/confluence/display/Hive/LanguageManual+WindowingAndAnalytics
+# * impala: https://docs.cloudera.com/documentation/enterprise/5-11-x/topics/impala_analytic_functions.html
+# * mssql: https://docs.microsoft.com/en-us/sql/t-sql/functions/first-value-transact-sql?view=sql-server-ver15
+# * oracle: https://oracle-base.com/articles/misc/first-value-and-last-value-analytic-functions
+# * redshift: https://docs.aws.amazon.com/redshift/latest/dg/r_WF_first_value.html
+# * teradata: https://docs.teradata.com/r/756LNiPSFdY~4JcCCcR5Cw/V~t1FC7orR6KCff~6EUeDQ
+#
+# no ignore nulls
+# * access: https://www.techonthenet.com/access/functions/
+# * hana: https://help.sap.com/viewer/4fe29514fd584807ac9f2a04f6754767/2.0.04/en-US/e7ef7cc478f14a408e1af27fc1a78624.html
+# * mysql: https://dev.mysql.com/doc/refman/8.0/en/window-function-descriptions.html
+
+#' Fill in missing values with previous or next value
+#'
+#' @inheritParams arrange.tbl_lazy
+#' @param ... Columns to fill.
+#' @param order_by Columns to order by before filling the values.
+#' @param .direction Direction in which to fill missing values. Currently
+#'   either "down" (the default) or "up".
+#'
+#' @export
+#'
+#' @examples
+#' squirrels <- tibble::tribble(
+#'   ~group,    ~name,     ~role,     ~n_squirrels, ~ n_squirrels2,
+#'   1,      "Sam",    "Observer",   NA,                 1,
+#'   1,     "Mara", "Scorekeeper",    8,                NA,
+#'   1,    "Jesse",    "Observer",   NA,                NA,
+#'   1,      "Tom",    "Observer",   NA,                 4,
+#'   2,     "Mike",    "Observer",   NA,                NA,
+#'   2,  "Rachael",    "Observer",   NA,                 6,
+#'   2,  "Sydekea", "Scorekeeper",   14,                NA,
+#'   2, "Gabriela",    "Observer",   NA,                NA,
+#'   3,  "Derrick",    "Observer",   NA,                NA,
+#'   3,     "Kara", "Scorekeeper",    9,                 10,
+#'   3,    "Emily",    "Observer",   NA,                NA,
+#'   3, "Danielle",    "Observer",   NA,                NA
+#' )
+#' squirrels$id <- 1:12
+#'
+#' dbplyr_fill(
+#'   tbl_memdb(squirrels),
+#'   n_squirrels,
+#'   n_squirrels2,
+#'   order_by = c(id)
+#' )
+dbplyr_fill <- function(.data, ..., order_by, .direction = c("down", "up")) {
+  # Sources
+  # * https://www.xspdf.com/resolution/51601218.html
+  # * https://dzone.com/articles/how-to-fill-sparse-data-with-the-previous-non-empt
+
+  sim_data <- simulate_vars(.data)
+  cols_to_fill <- syms(names(tidyselect::eval_select(expr(c(...)), sim_data)))
+  order_by_cols <- enquo(order_by)
+  .direction <- arg_match0(.direction, c("down", "up"))
+
+  if (.direction == "up") {
+    order_by_cols <- quo(-!!order_by_cols)
+  }
+
+  dbplyr_fill_int(
+    .con = remote_con(.data),
+    .data = .data,
+    cols_to_fill = cols_to_fill,
+    order_by_cols = order_by_cols,
+    .direction = .direction
+  )
+}
+
+dbplyr_fill_int <- function(.con, .data, cols_to_fill, order_by_cols, .direction) {
```

Comment:
Why the `_int` prefix?

Reply:
There is the user facing function `dbplyr_fill()`. It does some basic work and then it calls a separate function `dbplyr_fill_int()` that dispatches on the type of connection. I thought of the `_int` suffix as standing for "internal". Does this make sense to you or do you prefer a different name?

Reply:
Ohhhhhh, `_int` standards for `integer` to me, and we definitely want to avoid that. We don't have a strong existing convention for an internal suffix, but we been using 0 later (e.g. `dplyr_fill0()`)